File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    private int search(int[] nums,int target,boolean findStartIndex){
        int positionOfTarget=-1;
        int start=0;
        int end=nums.length-1;
        while(start<=end){
            int mid=start+(end-start)/2;
            if(target<nums[mid]) end=mid-1;
            else if(target>nums[mid]) start=mid+1;
            else{
                //potential position of the target is found but we have got to find the first and last position of it.
                positionOfTarget=mid;
                if(findStartIndex==true){
                    end=mid-1;
                }
                else {
                    start=mid+1;
                }
            }
        }
        return positionOfTarget;
    }
    public int[] searchRange(int[] nums, int target) {
        int[] firstAndLastPostionOfElementInSortedArray={-1,-1};
        //finding the first position of the target. 
        int firstPositionOfTarget=search(nums,target,true);
        //finding the last position of the target.
        int lastPositionOfTarget=search(nums,target,false);
        firstAndLastPostionOfElementInSortedArray[0]=firstPositionOfTarget;
        firstAndLastPostionOfElementInSortedArray[1]=lastPositionOfTarget;
        return firstAndLastPostionOfElementInSortedArray;
    }
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def upper(self,nums,target)->int:
        left=0
        right=len(nums)-1
        while left<=right:
            middle=left+(right-left)//2
            if nums[middle]<=target:
                left=middle+1
            else:
                right=middle-1
        if right==-1 or nums[right]!=target:
            return -1
        return right    
    def lower(self,nums,target)->int:
        left=0
        right=len(nums)-1
        while left<=right:
            middle=left+(right-left)//2
            if nums[middle]>=target:
                right=middle-1
            else:
                left=middle+1
        if left==len(nums) or nums[left]!=target:
            return -1
        return left
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        return [self.lower(nums,target),self.upper(nums,target)]
        